test/cpu_reduce_util.hpp

```diff
@@ -29,19 +29,194 @@
 #include <half.hpp>
 #include <limits>
 #include <cmath>
+#include <cassert>
 #include <miopen/miopen.h>
 #include <miopen/reduce_common.hpp>
 
 #include <miopen/bfloat16.hpp>
 
 namespace reduce {
 
+template <typename T>
+static inline bool IsNan(T x)
+{
+    // C++ isnan() is used for float and double
+    return (std::isnan(x));
+};
+
+template <>
+inline bool IsNan<half_float::half>(half_float::half x)
+{
+    return (half_float::isnan(x));
+};
+
+template <typename T>
+static inline bool IsFinite(T x)
+{
+    // C++ isfinite() is used for float and double
+    return (std::isfinite(x));
+};
+
+template <>
+inline bool IsFinite<half_float::half>(half_float::half x)
+{
+    return (half_float::isfinite(x));
+};
+
+template <typename T>
+static inline bool float_equal_one(T x)
+{
+    (void)x;
+    static_assert(static_cast<T>(0), "float_equal_one() is not implemented for this data type");
+    return false;
+};
+
+template <>
+inline bool float_equal_one<float>(float x)
+{
+    return std::isfinite(x) and x <= 1.0f and x >= 1.0f;
```

Comment:
[Q] Well, `(x == 1.0f)` should work properly and provides better performance. Why we need this?

Reply:
I don't how  x == const y  is implemented in instructions.  I used this <= and >= combination since it came across compiler warning previously (either for GPU or CPU kernel).

Reply:
[Recommendation] Add a comment explaining why the unusual code is needed to make it easier for readers to understand.

Reply:
The reason that I used the <= and >= combination is that this method removes the warning and should have better performance than using the std::nextafer by float_equal_fn defined in  src/include/miopen/float_equal.hpp

Reply:
@fronteer Is performance important?

[Hint] 
- You can use `x == 0.0` and `x==1.0` that should compile to a single instruction. I am not sure only about `half` on CPU.
- In case of insurmountable problems, bitwise comparison can be used, e.g. `(*reinterpret_cast<uint32_t*>(&x) == 0x3f800000)` for `float` etc.

Reply:
x == const y should be accurate if without compiler warning since here x is not used for calculated values, rather it is used for the constants to scaling the in/out tensor

Reply:
Thank you.   Changed.   Using == to compare floats on host side should not be an issue as x86 does have single instruction to do that and I saw half_float::half has "==" overloaded

Reply:
Just saw that  DeepCode report issue for using "==" with float/double

Reply:
This DeepCode warning is junk. It assumes that the programmer does not understand what he is doing.

Use this Ignore Suggestion from Deepcode if you care:

> Copy the following line:
> ```
> //  file deepcode ignore FloatingPointEquals: <comment the reason here>
> ```
> And paste it in cpu_reduce_util.hpp to prevent this suggestion from appearing again in this file. 

BTW the DeepCode's explanation is incorrect for zero, NaNz and Infinities:
> Floating point numbers are only considered equal if their bit representation is equal.

Reply:
I regressed to the previous codes and added some comments